src/web-auth/cross-origin-authentication.js

```diff
@@ -87,29 +99,47 @@ CrossOriginAuthentication.prototype.login = function(options, cb) {
           error: 'request_error',
           error_description: JSON.stringify(err)
         };
+
         return responseHandler(cb, { forceLegacyError: true })(errorObject);
       }
-      var popupMode = options.popup === true;
-      options = objectHelper.blacklist(options, [
-        'password',
-        'credentialType',
-        'otp',
-        'popup'
-      ]);
-      var authorizeOptions = objectHelper
-        .merge(options)
-        .with({ loginTicket: data.body.login_ticket });
-      var key = createKey(_this.baseOptions.rootUrl, data.body.co_id);
-      _this.storage.setItem(key, data.body.co_verifier, {
-        expires: times.MINUTES_15
-      });
-      if (popupMode) {
-        _this.webMessageHandler.run(
-          authorizeOptions,
-          responseHandler(cb, { forceLegacyError: true })
-        );
+
+      function doAuth() {
+        var popupMode = options.popup === true;
+
+        options = objectHelper.blacklist(options, [
+          'password',
+          'credentialType',
+          'otp',
+          'popup',
+          'onRedirecting'
+        ]);
+
+        var authorizeOptions = objectHelper
+          .merge(options)
+          .with({ loginTicket: data.body.login_ticket });
+
+        var key = createKey(_this.baseOptions.rootUrl, data.body.co_id);
+
+        _this.storage.setItem(key, data.body.co_verifier, {
+          expires: times.MINUTES_15
+        });
+
+        if (popupMode) {
+          _this.webMessageHandler.run(
+            authorizeOptions,
+            responseHandler(cb, { forceLegacyError: true })
+          );
+        } else {
+          _this.webAuth.authorize(authorizeOptions);
+        }
+      }
+
+      // Handle pre-redirecting to login, then proceed with '/authorize' once that is complete
+      if (typeof options.onRedirecting === 'function') {
+        options.onRedirecting(doAuth);
```

Comment:
Is this redirect called before hitting `/authorize` ? Does that mean authentication was not successful yet ? I guess not, because  we are only redirecting to `/authorize` to handle `consent` as the user is already logged in because of the `login_ticket`, right ?

Reply:
Yes, the hook is called in between getting the login ticket and redirecting to `/authorize`. It looks like successful authentication does not happen until the authz endpoint is called; just getting a valid login ticket is not enough.

Reply:
Keeping the original ESD in mind, does this solve what was actually being asked for? Being able to hook into successful logins?

Reply:
The intent of the work is to raise a hook in between getting a login ticket and calling the `/authorize` endpoint. I guess it depends on what is termed a "successful" login. I suppose if you _have_ a login ticket then the login could be considered successful, but Auth0 doesn't raise a "successful login" event until after the call to the authz endpoint.

I will however ask the customer to review the PR also before we merge to ensure it's what they want.

Reply:
@frederikprijck I asked the customer internally to review and they're happy it does the job 👍🏻